design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+    kind: ResourceConfigurationClass
+    name: gke-cluster
+  resourceRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: ResourceConfiguration
+    name: default-cluster-fdjk2
+---
+apiVersion: infrastructure.crossplane.io/v1alpha1
+kind: ResourceConfigurationClass
+metadata:
+  name: gke-cluster
+spec:
+  claim:
+    apiVersion: compute.example.org/v1alpha1
+    kind: KubernetesCluster
+  resources:
+  - base:
+      apiVersion: container.gcp.crossplane.io/v1beta1
+      kind: GKECluster
+      spec:
+        forProvider:
+          location: us-central1
+          initialClusterVersion: 1.16
+          masterAuth:
+            username: admin
+            clientCertificateConfig:
+              issueClientCertificate: true
+        writeConnectionSecretsToRef:
+          namespace: crossplane-system
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    patch:
+    - fromClaimField: ".metadata.uid"
+      toResourceField: ".spec.writeConnectionSecretToRef.name"
+    - fromClaimField: ".spec.version"
+      toResourceField: ".spec.forProvider.version"
+  - base:
+      apiVersion: container.gcp.crossplane.io/v1alpha3
+      kind: NodePool
+      # Resources cannot specify a name or a namespace (or rather, they are
+      # ignored if they are specified). The dynamic provisioning controller will
+      # derive provisioned resource names from their resource claim's name. This
+      # pattern exists today in Kubernetes, e.g. in a podTemplateSpec.
+      metadata:
+        annotations:
+          # Configuration classes that template more than one of a particular
+          # resource kind must set the external-name-prefix or suffix annotation
+          # in order to ensure multiple external resources of the same type are
+          # not created with the same name. These annotations will be set on the
+          # the provisioned managed resources and parsed by their controllers.
+          crossplane.io/external-name-suffix: "-b"
+      spec:
+        forProvider:
+          clusterSelector:
+            matchConfiguration: true
+          autoscaling:
+            enabled: true
+            minNodeCount: 0
+            maxNodecount: 100
+          config:
+            diskType: PD_SSD
+            machineType: n1-standard-8
+          version: 1.16
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    patch:
+    - fromClaimField: ".spec.version"
+      toResourceField: ".spec.forProvider.version"
+  # Note that the configuration class includes three node pool resources, two of
+  # which are identical. This design does not support a claim author requesting
+  # a configurable number of managed resources. This is intentional; doing so
+  # would blur the separation of concerns by shifting infrastructure concerns
+  # (how many node pools a cluster should have) away from the infrastructure
+  # owner and onto the application owner.
```

Comment:
Another possibly controversial area.

I think of a resource claim as the domain of the application owner, intended to describe an application's needs. _How_ the application's needs should be fulfilled is described in a resource class. You could imagine a resource claim allowing the author to configure how many and what shape node pools they get (as per the example in #152), but this starts to seem like the separation of concerns is being blurred - the resource claim author is describing _how_ their application's needs should be satisfied, rather than _what_ their application's needs are.

Reply:
What would you think about allowing for a resource count to be specified on the `ResourceConfigurationClass`? This would potentially the reduce the amount of duplicated YAML, but still keep all configuration on the side of the infrastructure owner.

Reply:
I'm assuming you mean adding a count that cannot be patched by a resource claim field? i.e. A count that would purely prevent resource class authors from having to duplicate YAML documents? If so, I would prefer to avoid it, primarily because introducing a `count` field would break (or at least complicate) the proposed external name prefix and suffix annotations. We'd produce N of the same kind of managed resource, all with the same external name.

Reply:
> this starts to seem like the separation of concerns is being blurred

Whenever I try to imagine a use case for this I keep coming back to something like `nodePools: 5` in a KubernetesCluster claim. Can folks think of some other use cases where this might be useful? I still feel like the "how many node pools do you want with this cluster" idea is leaking infrastructure concerns into the application-focused claim, but to be fair the whole idea of a `KubernetesCluster` claim goes somewhat against the idea of a claim being something that represents an application's needs.